### ISSUE TITLE
Director API POST - To update/add existing templates/variables

### ISSUE THREAD

**Issue by agconstantin** [Icinga/icingaweb2-module-director#1120]
Based on the director API doc, in order to update an existing objects, something like this can be used:
POST director/host?name=apitest
{
  "object_name": "apitest",
  "object_type": "object",
  "address": "127.0.0.1",
  "vars": {
    "location": "Berlin"
  }
}

I am using the following command:
./director-curl POST director/host?name=hostname '{ "object_name": "hostname", "address": "192.168.1.1", "object_type": "object", "imports": ["Environment PROD", "Environment DEV"] }'

With response:
HTTP/1.1 200 OK
Date: Fri, 25 Aug 2017 13:15:29 GMT
Server: Apache/2.4.6
X-Powered-By: PHP/5.4.16
Content-Length: 187
Content-Type: application/json

{
    "address": "192.168.1.1",
    "imports": [
        "Environment PROD",
        "Environment DEV"
    ],
    "object_name": "hostname",
    "object_type": "object"
}

But, if I want to update the object with new values, like for instance:
./director-curl POST director/host?name=hostname '{ "object_name": "hostname", "address": "192.168.1.1", "object_type": "object", "imports": ["Environment DEV", "Environment STAGING"] }'

The result would be:

{
    "address": "192.168.1.1",
    "imports": [
        "Environment DEV",
        "Environment STAGING"
    ],
    "object_name": "hostname",
    "object_type": "object"
}

## Expected Behavior

It should update the object parameters, not to overwrite them.

## Current Behavior

Overwrites current object parameters.

## Steps to Reproduce (for bugs)

1. Create host in icinga2 director (via API or not)
2. Execute POST on API address /director/host?name=hostname with additional parameters (eg. template)
3. Retrieve, GET on API address /director/host?name=hostname
4. Compare object data before step 2 with the one after step 3. Object data was not updated, it was just modified to match the provided parameters in step 2.

## Context

Updating Icinga2 Director object with new values.

## Your Environment

* Director version (System - About): 1.3.1
* Icinga Web 2 version and modules (System - About): 2.4.1
* Icinga 2 version (`icinga2 --version`): r2.7.0-1
* Operating System and version: Red Hat Enterprise Linux Server 7.3
* Webserver, PHP versions: Apache/2.4.6

**Comment by Thomas-Gelf** [comment]
@agconstantin: your example and the shown results look good to me. What exactly did you expect to happen? PUT overrides the whole object, POST changes only the given properties. As you pass in all used properties once again, the result looks correct to me.

**Comment by agconstantin** [comment]
Hi Thomas,
My object has these values:
{
"address": "192.168.1.1",
"imports": [
"Environment PROD",
"Environment DEV"
],
"object_name": "hostname",
"object_type": "object"
}

I run the POST to append/add a new value:
./director-curl POST director/host?name=hostname '{ "object_name": "hostname", "address": "192.168.1.1", "object_type": "object", "imports": ["Environment DEV", "Environment STAGING"] }'

Then my object becomes:
{
"address": "192.168.1.1",
"imports": [
"Environment DEV",
"Environment STAGING"
],
"object_name": "hostname",
"object_type": "object"
}

The Environment PROD has been removed.
Been a POST operation I would have expected only the append the newly values and not deleted the existing ones.

**Comment by Thomas-Gelf** [comment]
Ok, I understand. And no, this works as expected. `imports` can be seen as a property of type array. I guess most people would not expect to see `imports = ['a', 'b']` when posting `imports = ['a']` to an object carrying `imports = ['b']`. Same goes for `groups`. Also, please note that order matters for imports. So imagine two POSTs, one setting `imports = ['a']`, the other one `imports = ['b']`. Following your proposal, depending on your timing the result could be `a, b` or `b, a`. In theory we could add a feature allowing you to "add only missing members" via `"imports[]": "a"` or similar. But also this would sooner or later require you to write some code checking the current state first I guess.

Similar, but with more flexibility are `vars`. You can set `"vars": { "b": true }` - and `vars.a` will be lost. But there is the possibility to set just `"vars.b": true` in a POST, and `vars.a` would not be touched. But with `vars.b` already being a Custom Variable of type Array, it would be completely overwritten.

**Comment by agconstantin** [comment]
@Thomas-Gelf ,
Yes, I am doing such a logic in my script, to check whatever one of the host templates are attached to the host or not (in Director). If the host templates is not present, then I would like to add it.

So basically, what you are saying is that for now, the API does not support this on all level, like I am trying to do, is rather working fine for vars, but not for host templates, for instance?

Thanks.

**Comment by Thomas-Gelf** [comment]
No, I'm saying that such logic wouldn't make any sense and that it would be dangerous, as it wouldn't lead to a predictable state. What you have to do is to provide the `imports` property as it should look like after your POST - or leave it away, to not touch `imports` at all.

**Comment by agconstantin** [comment]
@Thomas-Gelf 
What are you saying, that a hosts in Icinga2 should have the same host templates, it had when it was initially bootstrapped? 
Well, what if you have existing hosts in Director, that you want to get them at the same level and configuration with the new ones? 

Let's take this scenario, you have a new service set attached to one and create a new host template for it as well, how do you attach this host template to all the hosts that you want?

**Comment by Thomas-Gelf** [comment]
@agconstantin: well, first of all I'd use an Import/Sync combination at that's much more powerful than all the scripts I've seen out there ;-) To add a template to a list of hosts I'd iterate through them, fetch the current `imports`property (you wrote that you already have such logic), insert the new template at the desired position (THIS is critical - order matters) - and then POST the new `imports` property.

**Comment by Thomas-Gelf** [comment]
NB: preferably I'd assign Service templates via assign rules based on Host properties and sync those properties. But there are probably things I do not know about in your environment forcing you to work the way you're doing it.

**Comment by Thomas-Gelf** [comment]
One more note: I'm not completely against adding a new feature allowing you to "append import members if not already set", but I'm completely against modifying the default behavior here as it would be just wrong.

So, the feature could look like this:

`POST ... '{..., "imports[]": ["new tpl"]}'`

The `[]` are saying "push to this array", like `+=`. Just, with some magic - if the template is already there, it would NOT be re-added to the end of the set. Here some sample hosts:

`
template Host "tpl a" { check_command = "hostalive" }
template Host "tpl b" { check_command = "hostalive6" }

object Host "a" {
  import "tpl a"
  import "tpl b"
}
object Host "b" {
  import "tpl b"
  import "tpl a"
}
object Host "c" {
}
`

Now "add" `tpl a` to all of them. What's the expected result?

**Comment by agconstantin** [comment]
@Thomas-Gelf ,

I was not aware the order matters, in that sense, when doing imports.
Secondly, in your last given example, the result should be, that `tpl a` is added to Host "c" and not to a or b, because it is already there.
For me, order of the templates, doesn't really make sense, because there is no dependency or order in how the checks attached to a template are run.
Lastly, yes, I am able to change my script and adjust it in such a way, that I can import a "standard" list of templates, that I need for that host, but I fear that if I will have custom hosts, that will have some "dedicated" host templates attached, in this way, I will override those custom ones, just that, otherwise I cannot complain too much..

**Comment by Thomas-Gelf** [comment]
Well... that's what I meant, you should always fetch the current object, merge your imports following your logic - and push the modified ones.

**Comment by agconstantin** [comment]
Yes and if there is a custom template, that was manually attached to that host, it will be removed in this way.

**Comment by Thomas-Gelf** [comment]
No, it will not - because you have fetched the current imports list from the Director API and merged it with the new template in your own code.

**Comment by Thomas-Gelf** [comment]
I could implement the `[]` feature I mentioned. Just wanted to make you aware of it's potential pitfalls. With your suggested result (I would also implement it that way) this would however mean that in that example one says "add 'tpl a' to all hosts", and might expect to have `hostalive` everywhere - while that's not the case. Host "a" would remain on `hostalive6`.

**Comment by agconstantin** [comment]
Yes and No, because we assume that you are aware of what you are doing and secondly, if something is not properly assigned where it is desired, you do have the option to delete it from there..

**Comment by Thomas-Gelf** [comment]
Who is *you*?

**Comment by agconstantin** [comment]
Initiator of the import/change, right? :)

**Comment by Thomas-Gelf** [comment]
Well... you're writing that tool as for some reason the existing Import/Sync implementation provided by Director isn't good enough. So, it's up to you to decide whether to write your tool in a defensive way or not. I'd strongly suggest to always assume that your users do not always know what they are doing, so you should protect them from themselves ;)

**Comment by agconstantin** [comment]
Hi @Thomas-Gelf ,

If you would like to consider this as a request for a feature would be nice, as I still consider the update procedure should work similar to how it is working an update into a database, where you do not have to dump whole set of data to reinsert it with the additional data ;)

Regards.

**Comment by Thomas-Gelf** [comment]
@agconstantin: you'll have the exact same problem when working with a database schema providing any kind of related _sorted_ set. You cannot "just add" a new member to that sorted set without first fetching the existing one. Anyways, sure, I'll consider this a feature request.
